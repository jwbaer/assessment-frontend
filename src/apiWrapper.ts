import axios from "axios";

const MAX_INFLIGHT_REQUESTS = 50;
const INTERVAL_MS = 1000;

let _currentInflightRequests = 0;

const API_URL = 'http://localhost:3030';

// this initial block sets up the axios client and
// all the needed interceptors

export const axiosInstance = axios.create();

// throttle to a max number of concurrent requests
// this helps to insure that in the case of a client side
// bug, we don't accidentally DDOS our own service
axiosInstance.interceptors.request.use((config) => {
    return new Promise((resolve, reject) => {
        let interval = setInterval(() => {
            if (_currentInflightRequests < MAX_INFLIGHT_REQUESTS) {
                _currentInflightRequests++;
                clearInterval(interval);
                resolve(config);
            }
        }, INTERVAL_MS);
    });
});

// some functions to handle waiting and retrying on a 429,
// in case the service has ingress controls that throttles us
// for some reason
const retryAfter = {
    isRetryable: (error: any) => {
        return !!(
            error.response &&
            error.response.status === 429 &&
            error.response.headers["retry-after"]
        );
    },

    wait: (error: any) => {
        return new Promise((resolve) =>
            setTimeout(resolve, error.response.headers["retry-after"]),
        );
    },

    retry: (error: any) => {
        if (!error.config) {
            throw error;
        } else {
            return axiosInstance(error.config);
        }
    },
};

// create the retry-after interceptor
axiosInstance.interceptors.response.use(null as any, async (error) => {
    if (retryAfter.isRetryable(error)) {
        await retryAfter.wait(error);
        return retryAfter.retry(error);
    } else {
        throw error;
    }
});

// once a response is done, make sure we decrement the inflight request count
axiosInstance.interceptors.response.use(
    (response) => {
        _currentInflightRequests = Math.max(0, _currentInflightRequests - 1);
        return Promise.resolve(response);
    },
    (error) => {
        if (
            error.response &&
            (error.response.status !== 429 ||
                !error.response.headers["retry-after"])
        ) {
            _currentInflightRequests = Math.max(
                0,
                _currentInflightRequests - 1,
            );
            return Promise.reject(error);
        }
    },
);

// the following methods provide an interface to call the APIs without having to
// scatter axios calls throughout the code.

export const receiveData = async (params: { url: string }) => {
    const endpoint = params.url;

    const response = await axiosInstance({
        method: "GET",
        url: endpoint,
        timeout: 60000,
    });
    if (response.status === 200) {
        return response.data;
    } else {
        throw new Error(response.data.message);
    }
};

// the API given for the assessment exercise is a read-only API,
// but this would be used by wrapper functions for any POST, PUT, PATCH,
// or DELETE calls, if they existed.  It's dead code at the moment,
// but I am including it for completeness, as I would typically make use
// of this in an api wrapper file such as this
export const sendData = async (
    params: { url: string; data?: any },
    method: any,
) => {
    const endpoint = params.url;

    const response = await axiosInstance({
        method,
        data: params.data,
        url: endpoint,
        timeout: 60000,
    });
    if (response.status === 200) {
        return response;
    } else {
        throw new Error(response.data.message);
    }
};

// pages
export const getPageSpec = (pageId: string) => {
    const url = `${API_URL}/page/${pageId}`;
    return receiveData({ url });
};

// weather
export const getWeather = (lat: number, lon: number) => {
    const url = `${API_URL}/weather/?lat=${lat}&lon=${lon}`;
    return receiveData({ url });
};
