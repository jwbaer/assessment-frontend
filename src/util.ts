import { Component, List, PageApiResult, PageSpec, Variable, VariableDeclaration } from "./types";

export const convertPageApiResultToPageSpec = (apiResult: PageApiResult): PageSpec => {
    const componentIndex: Map<number, Component> = new Map(
        apiResult.components.map((obj: Component) => [obj.id, obj])
    );
    const listIndex: Map<number, List> = new Map(
        apiResult.lists.map((obj: List) => [obj.id, obj])
    );
    const variableIndex: Map<string, Variable> | undefined = apiResult.variables
        ? new Map(
            apiResult.variables.map((obj: VariableDeclaration) => [
                obj.name,
                { name: obj.name, type: obj.type, value: obj.initialValue },
            ])
        )
        : undefined;

    return({ componentIndex, listIndex, variableIndex });
}
