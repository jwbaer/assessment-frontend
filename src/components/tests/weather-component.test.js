import * as apiWrapper from "../../apiWrapper";

import { act, render, screen, waitFor } from "@testing-library/react";

import ComponentPanel from "../component-panel";
import PageContext from "../../contexts/page-context";
import WeatherComponent from "../weather-component";
import { convertPageApiResultToPageSpec } from "../../util";

jest.mock("../../apiWrapper");

describe("<WeatherComponent /> spec", () => {
    beforeEach(() => jest.clearAllMocks());

    const apiResult = {
        lists: [
            {
                id: 0,
                components: [1],
            },
        ],
        components: [
            {
                id: 1,
                type: "weather",
                options: {
                    lon: "40.748607102729295",
                    lat: "-73.98563758004718",
                },
            },
        ],
    };

    const data = {
        lon: "40.748607102729295",
        lat: "-73.98563758004718",
        condition: "cloudy",
        conditionName: "Cloudy",
        temperature: 78,
        unit: "f",
        location: "New York, NY",
        upcomming: [
            {
                day: "Fri",
                condition: "cloudy",
                conditionName: "Cloudy",
            },
            {
                day: "Sat",
                condition: "cloudy",
                conditionName: "Cloudy",
            },
            {
                day: "Sun",
                condition: "rain",
                conditionName: "Rain",
            },
        ],
    };

    const pageSpec = convertPageApiResultToPageSpec(apiResult);

    it("renders the component", async () => {
        apiWrapper.getWeather.mockResolvedValue({ data });
        const { asFragment } = render(
            <WeatherComponent
                component={{
                    id: 1,
                    type: "weather",
                    options: {
                        lon: "40.748607102729295",
                        lat: "-73.98563758004718",
                    },
                }}
            />
        );

        expect(screen.getByText("Loading...")).toBeDefined();

        await act(
            async () =>
                await waitFor(() => {
                    expect(screen.getByText("New York, NY")).toBeDefined();
                    expect(screen.getByText(/78.F/i)).toBeDefined();
                    expect(screen.getByText("Cloudy")).toBeDefined();
                    expect(screen.getByText("Fri")).toBeDefined();
                    expect(screen.getByText("Sat")).toBeDefined();
                    expect(screen.getByText("Sun")).toBeDefined();
                    expect(asFragment()).toMatchSnapshot();
                })
        );
    });

    it("renders an error on a API error", async () => {
        apiWrapper.getWeather.mockRejectedValue({});
        const { asFragment } = render(
            <WeatherComponent
                component={{
                    id: 1,
                    type: "weather",
                    options: {
                        lon: "40.748607102729295",
                        lat: "-73.98563758004718",
                    },
                }}
            />
        );

        expect(screen.getByText("Loading...")).toBeDefined();

        await act(
            async () =>
                await waitFor(() => {
                    expect(screen.getByText("Error loading weather")).toBeDefined();
                    expect(asFragment()).toMatchSnapshot();
                })
        );
    });

    it("renders when specifed through the component panel", async () => {
        apiWrapper.getWeather.mockResolvedValue({
            data,
        });

        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec }}>
                <ComponentPanel componentId={1} />
            </PageContext.Provider>
        );

        expect(screen.getByText("Loading...")).toBeDefined();

        await act(
            async () =>
                await waitFor(() => {
                    expect(screen.getByText("New York, NY")).toBeDefined();
                    expect(screen.getByText(/78.F/i)).toBeDefined();
                    expect(screen.getByText("Cloudy")).toBeDefined();
                    expect(screen.getByText("Fri")).toBeDefined();
                    expect(screen.getByText("Sat")).toBeDefined();
                    expect(screen.getByText("Sun")).toBeDefined();
                    expect(asFragment()).toMatchSnapshot();
                })
        );
    });
});
