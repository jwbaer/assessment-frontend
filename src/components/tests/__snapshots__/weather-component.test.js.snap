// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherComponent /> spec renders an error on a API error 1`] = `
<DocumentFragment>
  <div
    class="component weather placeholder"
  >
    Error loading weather
  </div>
</DocumentFragment>
`;

exports[`<WeatherComponent /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="component weather"
    data-testid="weather-component-1"
  >
    <div
      class="top-row"
    >
      <div
        class="current-weather"
      >
        <div
          class="main-weather-icon"
        >
          <svg
            height="128"
            viewBox="0 0 128 128"
            width="128"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Clouds
            </title>
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h128v128H0z"
                fill="#FFF"
              />
              <path
                d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
                fill="#DDE1E6"
                fill-opacity="0.8"
              />
            </g>
          </svg>
        </div>
        <div
          class="current-temp-and-condition"
        >
          <div
            class="current-temp"
          >
            78°F
          </div>
          <div
            class="current-condition"
          >
            Cloudy
          </div>
        </div>
      </div>
      <div
        class="location"
      >
        New York, NY
      </div>
    </div>
    <div
      class="bottom-row"
    >
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Clouds
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
              fill-opacity="0.8"
            />
          </g>
        </svg>
        Fri
      </div>
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Clouds
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
              fill-opacity="0.8"
            />
          </g>
        </svg>
        Sat
      </div>
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Rain
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 85.974C32.772 85.454 25 77.158 25 67c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 40.48 58.97 34 67.9 34c9.904 0 17.993 7.971 18.5 18h.025C95.579 52 103 59.611 103 69s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
            />
            <path
              d="M77 92c0-1.105.888-2 2-2 1.105 0 2 .887 2 2v6c0 1.105-.888 2-2 2-1.105 0-2-.887-2-2v-6zm-10-.006c0-1.101.888-1.994 2-1.994 1.105 0 2 .895 2 1.994v12.012A1.992 1.992 0 0169 106c-1.105 0-2-.895-2-1.994V91.994zM57 92c0-1.105.888-2 2-2 1.105 0 2 .887 2 2v6c0 1.105-.888 2-2 2-1.105 0-2-.887-2-2v-6zm-10-.006c0-1.101.888-1.994 2-1.994 1.105 0 2 .895 2 1.994v12.012A1.992 1.992 0 0149 106c-1.105 0-2-.895-2-1.994V91.994z"
              fill="#1994FF"
            />
          </g>
        </svg>
        Sun
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<WeatherComponent /> spec renders when specifed through the component panel 1`] = `
<DocumentFragment>
  <div
    class="component weather"
    data-testid="weather-component-1"
  >
    <div
      class="top-row"
    >
      <div
        class="current-weather"
      >
        <div
          class="main-weather-icon"
        >
          <svg
            height="128"
            viewBox="0 0 128 128"
            width="128"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Clouds
            </title>
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h128v128H0z"
                fill="#FFF"
              />
              <path
                d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
                fill="#DDE1E6"
                fill-opacity="0.8"
              />
            </g>
          </svg>
        </div>
        <div
          class="current-temp-and-condition"
        >
          <div
            class="current-temp"
          >
            78°F
          </div>
          <div
            class="current-condition"
          >
            Cloudy
          </div>
        </div>
      </div>
      <div
        class="location"
      >
        New York, NY
      </div>
    </div>
    <div
      class="bottom-row"
    >
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Clouds
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
              fill-opacity="0.8"
            />
          </g>
        </svg>
        Fri
      </div>
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Clouds
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 89.974C32.772 89.454 25 81.158 25 71c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 44.48 58.97 38 67.9 38c9.904 0 17.993 7.971 18.5 18h.025C95.579 56 103 63.611 103 73s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
              fill-opacity="0.8"
            />
          </g>
        </svg>
        Sat
      </div>
      <div
        class="weather-icon"
      >
        <svg
          height="128"
          viewBox="0 0 128 128"
          width="128"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Rain
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="#FFF"
            />
            <path
              d="M42.55 85.974C32.772 85.454 25 77.158 25 67c0-10.493 8.294-19 18.525-19 2.19 0 4.29.39 6.24 1.105C51.517 40.48 58.97 34 67.9 34c9.904 0 17.993 7.971 18.5 18h.025C95.579 52 103 59.611 103 69s-7.42 17-16.575 17H42.55v-.026z"
              fill="#DDE1E6"
            />
            <path
              d="M77 92c0-1.105.888-2 2-2 1.105 0 2 .887 2 2v6c0 1.105-.888 2-2 2-1.105 0-2-.887-2-2v-6zm-10-.006c0-1.101.888-1.994 2-1.994 1.105 0 2 .895 2 1.994v12.012A1.992 1.992 0 0169 106c-1.105 0-2-.895-2-1.994V91.994zM57 92c0-1.105.888-2 2-2 1.105 0 2 .887 2 2v6c0 1.105-.888 2-2 2-1.105 0-2-.887-2-2v-6zm-10-.006c0-1.101.888-1.994 2-1.994 1.105 0 2 .895 2 1.994v12.012A1.992 1.992 0 0149 106c-1.105 0-2-.895-2-1.994V91.994z"
              fill="#1994FF"
            />
          </g>
        </svg>
        Sun
      </div>
    </div>
  </div>
</DocumentFragment>
`;
